src/main/scala/org/broadinstitute/dsde/workbench/sam/azure/AzureService.scala

```diff
@@ -126,6 +133,70 @@ class AzureService(crlService: CrlService, directoryDAO: DirectoryDAO, azureMana
       createdPet <- directoryDAO.createPetManagedIdentity(petToCreate, samRequestContext)
     } yield (createdPet, true)
 
+  def getOrCreateActionManagedIdentity(
+      resource: FullyQualifiedResourceId,
+      resourceAction: ResourceAction,
+      mrgCoordinates: ManagedResourceGroupCoordinates,
+      samUser: SamUser,
+      samRequestContext: SamRequestContext
+  ): IO[(ActionManagedIdentity, Boolean)] = {
+    val id = ActionManagedIdentityId(resource, resourceAction, mrgCoordinates)
+    for {
+      hasPermission <- policyEvaluatorService.hasPermission(resource, resourceAction, samUser.id, samRequestContext)
```

Comment:
Usually permission checks happen in routes, see the `requireOneOfAction` directive. The reasons being that a) there is a common place for permission checks to happen, b) the service code is more reusable and c) common handling of 404 vs 403 responses

Reply:
I've recently been questioning this approach. I'm not sure which is better: doing a permission check at the "point of ingress" or the "point of action". 

Doing it at the point of ingress means that the service code can be simpler, but it opens up possibilities for developer error. Because the check and the action are not co-located, it becomes easy to call the service code from another place in the codebase, maybe a place where the developer didn't or forgot to implement a permissions check. 

By co-locating the permissions check with the action in the service code, it takes that burden off the developer. A developer no longer needs to carry that "context" as they develop: they can be sure that the service code won't let the user do something they're not allowed to do. However, it does make the service code a bit more complicated. 

I'm not sure which approach is better. Maybe neither is. Interested in hearing your thoughts.

Reply:
I moved everything to the routes layer